internal/tests/cluster/session_cleanup_test.go

```diff
@@ -0,0 +1,513 @@
+
+	// Run initial send/receive test, make sure things are working
+	sConn.TestSendRecvAll(t)
+
+	// Kill connection to first controller, and run test again, should
+	// pass, deferring to other controller.
+	p1.Pause()
+	sConn.TestSendRecvAll(t)
+
+	// Resume first controller, pause second. This one should work too.
+	p1.Resume()
+	p2.Pause()
+	sConn.TestSendRecvAll(t)
+
+	// Kill the first controller connection again. This one should fail
+	// due to lack of any connection.
+	p1.Pause()
+	sConn.TestSendRecvFail(t)
+
+	// Finally resume both, try again. Should behave as per normal.
+	p1.Resume()
+	p2.Resume()
+	time.Sleep(time.Second * 10)                          // Sleep to wait for worker to report back as healthy
+	sess = newTestSession(ctx, t, tcl, "ttcp_1234567890") // re-assign, other connection will close in t.Cleanup()
+	sConn = sess.Connect(ctx, t, logger)
+	sConn.TestSendRecvAll(t)
+}
+
+// testSession represents an authorized session.
+type testSession struct {
+	workerAddr      string
+	transport       *http.Transport
+	tofuToken       string
+	connectionsLeft int32
+}
+
+// newTestSession authorizes a session and creates all of the data
+// necessary to initialize.
+func newTestSession(
+	ctx context.Context,
+	t *testing.T,
+	tcl *targets.Client,
+	targetId string,
+) *testSession {
+	require := require.New(t)
+	sar, err := tcl.AuthorizeSession(ctx, "ttcp_1234567890")
+	require.NoError(err)
+	require.NotNil(sar)
+
+	s := new(testSession)
+	authzString := sar.GetItem().(*targets.SessionAuthorization).AuthorizationToken
+	marshaled, err := base58.FastBase58Decoding(authzString)
+	require.NoError(err)
+	require.NotZero(marshaled)
+
+	sessionAuthzData := new(targetspb.SessionAuthorizationData)
+	err = proto.Unmarshal(marshaled, sessionAuthzData)
+	require.NoError(err)
+	require.NotZero(sessionAuthzData.GetWorkerInfo())
+
+	s.workerAddr = sessionAuthzData.GetWorkerInfo()[0].GetAddress()
+
+	parsedCert, err := x509.ParseCertificate(sessionAuthzData.Certificate)
+	require.NoError(err)
+	require.Len(parsedCert.DNSNames, 1)
+
+	certPool := x509.NewCertPool()
+	certPool.AddCert(parsedCert)
+	tlsConf := &tls.Config{
+		Certificates: []tls.Certificate{
+			{
+				Certificate: [][]byte{sessionAuthzData.Certificate},
+				PrivateKey:  ed25519.PrivateKey(sessionAuthzData.PrivateKey),
+				Leaf:        parsedCert,
+			},
+		},
+		RootCAs:    certPool,
+		ServerName: parsedCert.DNSNames[0],
+		MinVersion: tls.VersionTLS13,
+	}
+
+	s.transport = cleanhttp.DefaultTransport()
+	s.transport.DisableKeepAlives = false
+	s.transport.TLSClientConfig = tlsConf
+	s.transport.IdleConnTimeout = 0
+
+	return s
+}
+
+// connect returns a connected websocket for the stored session,
+// connecting to the stored workerAddr with the configured transport.
+//
+// The returned (wrapped) net.Conn should be ready for communication.
+func (s *testSession) connect(ctx context.Context, t *testing.T) net.Conn {
+	require := require.New(t)
```

Comment:
should we add a call to t.Helper() here?

Reply:
This should be addressed now!